Implement URL updates and browser history for grants table interactions

jeffsmohan opened the pull request:
### Ticket #2575
## Description
**Note: I wanted some early eyes on this code change to make sure it seems directionally correct. I'd still like us to do more manual testing on the feature, as I worry there are still outstanding edge cases or interactions e.g. with saved searches that I don't have full product context on.**

Implements browser history and URL updates when interacting with the grants table. The interactions are now represented in a URL query param and added to history (i.e., browser back/forward will take you to those states): 

- Pagination (going to page 2, 3, etc.) `?page=2`
- Sorting (sort a column ascending or descending) `?sort=close_date` or `?sort=close_date&desc=true`
- Saved search (you've searched/filtered the results) `?search=1234`
- Any combination of the above `?search=1234&sort=close_date&desc=true&page=2`

Additionally, adds the active tab into the URL when switching between interested, assigned, not applying, and applied sub-tabs on the My Grants tab: 

- My Grants sub-tab `/my-grants/interested` or `/my-grants/not-applying`
- Combinations with the table query params above `/my-grants/not-applying?search=1234&sort=close_date&desc=true&page=2`

By adding these as entries in the browser history, the "back" functionality from he Grants Detail page should now work as expected, taking you right back to where you were and ensuring the specific grant you were looking at is still listed where you left it in the grant table when you return. 

## Screenshots / Demo Video
<img width="822" alt="image" src="https://github.com/usdigitalresponse/usdr-gost/assets/126250/d0cd8df5-726a-4012-a02d-a860cc53cbb1">

## Testing
It can be very helpful to have a larger and more realistic dataset loaded up for testing sorting/searching/pagination. There may be a better way, but here's how I got a dump of staging data into my local environment: 

1. Download a staging db dump and unzip it
    - e.g., https://usdigitalresponse.slack.com/files/U042RBLELUT/F06A9P2KLFM/gost-dump-full_2023-12-12.sql.tar.gz
    - Slack reference: https://usdigitalresponse.slack.com/archives/C0324KDQSCR/p1707931613492139
2. `docker compose down` (to start from a fresh Postgres install)
3. `docker compose up -d`
4. `docker compose run -v ~/Downloads:/downloads app psql -d postgresql://postgres:password123@postgres:5432/usdr_grants -f /downloads/dump.sql`
    - Point the `-v` option to where your unzipped dump is located; point the `-f` option to the name of the dump
5. `docker compose exec app yarn db:migrate`

Once you have that, you can verify the URL gets updated as you interact with the grants tables throughout the site, and verify that the browser back/forward behavior works as expected. 

### Automated and Unit Tests
- [ ] Added Unit tests

### Manual tests for Reviewer
- [ ] Added steps to test feature/functionality manually

## Checklist
- [ ] Provided ticket and description
- [ ] Provided screenshots/demo
- [ ] Provided testing information
- [ ] Provided adequate test coverage for all new code
- [ ] Added PR reviewers

inaki commented:
It looks good to me. 👍 
- tested sort and search and url is representing the events
- tested browser b/f navigation 
- tested url pasting and get into the correct state

jeffsmohan commented:
@inaki I fixed an issue where an initial page load with a search id AND sort params would reset the sorting. Also merged main. Please re-review!

jeffsmohan commented:
@inaki Ok, sorry for the delay, but I think this PR is ready for re-review! To recap, the changes since your last review: 

1. Fixed an issue where an initial page load with a search id AND sort params would reset the sorting
2. Improved grants api endpoint to return 400 for invalid pagination params (instead of 500ing, for example, on `?page=-1`)
3. Clamps page to valid interval once grants have been fetched (e.g., if you saved a URL like `?search=123&page=5`, which used to return 5 pages but now only has 4 pages of results because of some grants expiring or whatever, it'll now restrict that to a valid page number)
4. Removes search query if it's not found (e.g., if you saved a URL like `?search=1234` and shared it, but they don't have a saved search with ID 1234, it'll reset to null search id.)
5. Fixed a bug where clearing the active saved search by hitting the "clear" link wasn't working

(These map cleanly to individual commits, so that may be the easiest way to review.)

sanason commented:
I found one issue while clicking around:

1. Click on "My Grants" tab. URL is `/my-grants/interested`.
2. Click on a row in the grants table (assuming you have at least one "Interested" grant).
**Expected:** URL is `/grant/:id` and Grant Details page is displayed.
**Actual:** URL is `/my-grants/grant/:id` and 404 page is displayed.

sanason reviewed (commented):
Works well, just a couple of issues I found from clicking around.

One question: Did we decide, for this ticket, that it's not important to have shareable search URLs?

sanason reviewed (approved):
As far as I can tell, this works well.

> I think this points to a broader concern I have with the overall architecture of this, given how fiddly and difficult to reason about it seems to have become. 

I agree that this has gotten hard to reason about (no fault of your own). Personally, I'm not a big fan of global state on the client side so I'd tend to blame the complexity on too much Vuex and not enough local state + props.

> But that's a broader issue that hopefully I can address in some future work.

+1

inaki commented:
It looks good to me 👍

jeffsmohan commented:
> Personally, I'm not a big fan of global state on the client side

@sanason oh my goodness, 💯 agree!
